mule-extensions-api/src/main/java/org/mule/runtime/extension/api/introspection/AbstractComplexModel.java

```diff
@@ -93,32 +100,53 @@ public AbstractComplexModel(String name,
    */
   @Override
   public List<ConnectionProviderModel> getConnectionProviders() {
-    return toList(connectionProviders.values());
+    return connectionProviders;
   }
 
-  protected <T extends EnrichableModel> Optional<T> findModel(Map<String, T> map, String name) {
-    return Optional.ofNullable(map.get(name));
+  /**
+   * Returns the first item in the {@code values} collection which
+   * matches the given {@code name}.
+   *
+   * @param values a {@link Collection} of {@link Named} items
+   * @param name   the matching criteria
+   * @param <T>    the generic type of the {@code values} items
+   * @return an {@link Optional} matching item
+   */
+  protected <T extends Named> Optional<T> findModel(Collection<T> values, String name) {
+    return values.stream().filter(v -> v.getName().equals(name)).findFirst();
   }
 
   protected <T extends Described> List<T> toList(Collection<T> collection) {
     if (collection == null || collection.isEmpty()) {
-      return Collections.emptyList();
+      return emptyList();
     }
-    return Collections.unmodifiableList(new ArrayList<>(collection));
+    return unmodifiableList(new ArrayList<>(collection));
   }
 
-  protected <T extends Named> Map<String, T> toMap(List<T> objects) {
-    if (objects == null || objects.isEmpty()) {
-      return Collections.emptyMap();
-    }
+  /**
+   * Returns an immutable copy of the {@code values} collection, validating
+   * that no items exist such that its name is repeated
+   *
+   * @param values     the collection to copy
+   * @param identifier human friendly identifier of the {@code values} content
+   * @param <T>        the generic type of the {@code values} items
+   * @return an immutable copy of the {@code values}
+   */
+  protected <T extends Named> List<T> unique(Collection<T> values, String identifier) {
+    Multiset<String> names = HashMultiset.create();
+    values.stream().map(Named::getName).forEach(names::add);
 
-    Map<String, T> map = new LinkedHashMap<>(objects.size());
-    for (T object : objects) {
-      if (map.containsKey(object.getName())) {
-        throw new IllegalArgumentException(String.format("Multiple entries with the same key[%s]", object.getName()));
-      }
-      map.put(object.getName(), object);
+    List<String> invalid = names.entrySet().stream()
+        .filter(entry -> entry.getCount() > 1)
+        .map(Multiset.Entry::getElement)
+        .collect(Collectors.toList());
+
+    if (!invalid.isEmpty()) {
+      throw new IllegalModelDefinitionException(format("%s [%s] were defined multiple times",
+                                                       identifier,
+                                                       Joiner.on(", ").join(invalid)));
```

Comment:
There is no need for use the `Joiner`, the String list will be printed correctly